src/ray/core_worker/transport/out_of_order_actor_scheduling_queue.cc

```diff
@@ -69,31 +74,54 @@ void OutOfOrderActorSchedulingQueue::Add(
     const std::string &concurrency_group_name,
     const ray::FunctionDescriptor &function_descriptor,
     TaskID task_id,
+    uint64_t attempt_number,
     const std::vector<rpc::ObjectReference> &dependencies) {
   RAY_CHECK(boost::this_thread::get_id() == main_thread_id_);
   auto request = InboundRequest(std::move(accept_request),
                                 std::move(reject_request),
                                 std::move(send_reply_callback),
                                 task_id,
-                                dependencies.size() > 0,
+                                attempt_number,
+                                dependencies,
                                 concurrency_group_name,
                                 function_descriptor);
+  bool run_request = true;
+  std::optional<InboundRequest> request_to_cancel;
   {
     absl::MutexLock lock(&mu_);
-    pending_task_id_to_is_canceled.emplace(task_id, false);
+    if (pending_task_id_to_is_canceled.contains(task_id)) {
+      // There is a previous attempt of the same task running,
```

Comment:
it's not solid to test "there's task running" by `pending_task_id_to_is_canceled.contains`.

1. Task `Add()`ed, call `pending_task_id_to_is_canceled.emplace()`
2. `AcceptRequestOrRejectIfCanceled()` called after deps are ready. In which:
3. Task starts running by `request.Accept()` 
4. `pending_task_id_to_is_canceled.erase()`ed
5. (async) Task finished, calling rpc callback

in 2-4 the task is pending but not running. in 4-5, the task is running, but its task_id is already erased from pending_task_id_to_is_canceled. 

I think the goal is:

1. if a prev attempt is already running, queue this new attempt
2. if a prev attempt is pending, fail the prev attempt in favor of the new attempt
3. if there's no prev attempt (never existed or finished running), just run

so we need to track the state of each attempt run (Pending/Pending-cancelled/Running/NA) instead of only relying on the existing states (Pending/Pending-cancelled/NA)

Reply:
`Accept()` is sync so when it returns, the task execution is done.